File: kotlin-archive/jam-core/src/main/kotlin/io/forge/jam/core/Hashing.kt
package io.forge.jam.core

import org.bouncycastle.crypto.digests.KeccakDigest
import org.bouncycastle.jcajce.provider.digest.Blake2b

/**
 * Blake2b-256 hash function.
 */
fun blakeHash(data: ByteArray): ByteArray {
    val digest = Blake2b.Blake2b256()
    digest.update(data, 0, data.size)
    return digest.digest()
}

/**
 * Keccak-256 hash function.
 */
fun keccakHash(data: ByteArray): ByteArray {
    val digest = KeccakDigest(256)
    val output = ByteArray(32)
    digest.update(data, 0, data.size)
    digest.doFinal(output, 0)
    return output
}

File: modules/core/src/main/scala/io/forge/jam/core/hashing.scala
package io.forge.jam.core

import org.bouncycastle.jcajce.provider.digest.Blake2b
import org.bouncycastle.crypto.digests.KeccakDigest
import primitives.Hash

/**
 * Hashing utilities for JAM protocol.
 *
 * Provides Blake2b-256 and Keccak-256 hash functions using Bouncy Castle.
 */
object Hashing:

  /**
   * Compute Blake2b-256 hash of the given data.
   *
   * @param data The input data as JamBytes
   * @return The 32-byte hash as Hash type
   */
  def blake2b256(data: JamBytes): Hash =
    blake2b256(data.toArray)

  /**
   * Compute Blake2b-256 hash of the given data.
   *
   * @param data The input data as byte array
   * @return The 32-byte hash as Hash type
   */
  def blake2b256(data: Array[Byte]): Hash =
    val digest = new Blake2b.Blake2b256()
    digest.update(data, 0, data.length)
    Hash(digest.digest())

  /**
   * Compute Keccak-256 hash of the given data.
   *
   * @param data The input data as JamBytes
   * @return The 32-byte hash as Hash type
   */
  def keccak256(data: JamBytes): Hash =
    keccak256(data.toArray)

  /**
   * Compute Keccak-256 hash of the given data.
   *
   * @param data The input data as byte array
   * @return The 32-byte hash as Hash type
   */
  def keccak256(data: Array[Byte]): Hash =
    val digest = new KeccakDigest(256)
    val output = new Array[Byte](32)
    digest.update(data, 0, data.length)
    digest.doFinal(output, 0)
    Hash(output)
